pkg/storage/bucket/s3/bucket_client.go

```diff
@@ -153,8 +154,18 @@ func (b *BucketWithRetries) Exists(ctx context.Context, name string) (exists boo
 }
 
 func (b *BucketWithRetries) Upload(ctx context.Context, name string, r io.Reader) error {
+	// Convert Reader to Seeker
+	var buf bytes.Buffer
+	if _, err := buf.ReadFrom(r); err != nil {
+		return err
+	}
+	s := bytes.NewReader(buf.Bytes())
 	return b.retry(ctx, func() error {
-		return b.bucket.Upload(ctx, name, r)
+		err := b.bucket.Upload(ctx, name, s)
```

Comment:
If `err != nil` here, should we return `err` early instead of going ahead and reset the byte slice reader to the start? If the intention is to always reset, then I think code code may read better if we do the seek before the upload; it signals that seeking to the start is a pre-condition for upload to be successful.